MI-209 MI-210 MI-211 Key Value Storage and Data Layer

limphilip opened the pull request:
This pull request contains the following changes:

- The server logic from the Mappers has been moved to a new package: `Gateways`, which is responsible for dealing with logic that requires communication components outside the boundaries of the application. For example, communication with the web API. According to the SAD, the `Gateways` package is found under the data layer.
- A new package named `Cache` has been created. It contains for now, a `UserCache` that is responsible for dealing with local persistent storage related to the User model. For now, the key-value pair to save the user's preference is present.
- Upon creating a user, the new user's preferences are saved locally on the persistent storage using the `UserCache`. This is done by the `UserMapper` which now have a new attribute: `userCache`.
- Unit tests have been added to test the Key-Value pair storage for the user's preferences.
- Also affected, the mappers in the `UnitTest` project have been changed to `Gateways` as well.
- Minor changes for variable names have been performed while refactoring the data layer for consistency.

Also resolved many refactoring problems of #24

limphilip committed:
MI-209 Addition of a data layer, moved server logic from mapper to gateways

In order to have key value pair implementation, there must be a data
layer, currently absent. This commit contains the new data layer, and
the server logic originally found in the mappers have been moved to the
gateway package (Gateways namespace). The mappers now uses the gateway
to retrieve information from the web.

For now, the data layer may seem useless, but once local persistent
storage is implemented, the mapper first looks in the local persistent
storage for information, and only if it is not found locally, then it
will use the gateways to find the information. The Key value pairs will
be part of the local storage.

limphilip committed:
MI-210 Initial implementation of user cache and MI-211 initial tests for user cache

MI-210 Implements the user cache containing methods that takes advantage
of key-value pair persistent storage. It allows writing and reading user
preferences from the persistent storage.

MI-211 Unit tests for writing and reading user preferences from
persistent storage using the new user cache

In the unit test project, replaced mapper with gateway as per the new
data layer on the app project.

limphilip committed:
MI-209 Csproj file: corrected the paths to the new gateways

After changing the mappers to gateways, the paths were broken in the
csproj which makes the project unable to load. Fixed that.

limphilip committed:
MI-211 Updated unit test for user cache

Updated unit test for user cache to verify more of the data retrieved
for user preferences.

Also corrected android manifest from cap letter to lower case letter to
avoid deployment error.

Updated the link mode in android for unit test csproj file to be able to
compile.

limphilip committed:
MI-210 Save user's preference when creating a new user

The UserMapper has been updated to save the user's preferences when a
new user is created. The UserMapper makes use of its new userCache
attribute to do so.

Ideawin commented:
I noticed that you started using the "summary" comments for documentation in UserGateway only, but you didn't put them in your other classes like AbstractGateway.cs and ConsoleGateway.cs. Please don't forget to add them.

limphilip committed:
MI-210 Added XML comments to the AbstractGateway and ConsoleGateway

The comments were in an old format, not following in XML convention.
They have been changed, updated and added to the AbstractGateway and
ConsoleGateway.

limphilip commented:
Thanks @Ideawin ! I have converted the old comments to the XML format, and added missing comments on those files.

Ideawin commented:
Perfect, everything seems good. Merging :)
